### PR TITLE
MAINT Use run_in_pyodide in xgboost, opencv, and matplotlib tests

### PR DESCRIPTION
**Pull request by ryanking13**
This PR refactors tests in xgboost, opencv-python, and matplotlib to use `run_in_pyodide`. No other things are affected

### CODE REVIEW

File: packages/opencv-python/test_opencv_python.py [hunk 1]
```diff
@@ -1,42 +1,42 @@
...
+            )
+            pixels_total = image.size
 
-    # Due to some randomness in the result, we allow a small difference
-    return match_ratio > 0.95
+            match_ratio = float(pixels_match / pixels_total)
+            assert match_ratio > MIN_MATCH_RATIO
 
+        # TODO: this is a hack to make the function available in the global scope
+        np.compare_ = compare
```
Comment (ryanking13) [inline packages/opencv-python/test_opencv_python.py:37]: I wasn't able to find a better way to make a function or variable defined in some `run_in_pyodide` block visible in another `run_in_pyodide` block. Do you have any suggestions @hoodmane?

Comment (hoodmane) [reply packages/opencv-python/test_opencv_python.py:37]: Well I got a bit nerd sniped by this but I think I have a solution.

Comment (hoodmane) [reply packages/opencv-python/test_opencv_python.py:37]: See PR #49.

File: packages/opencv-python/test_opencv_python.py [hunk 2]
```diff
@@ -137,21 +141,21 @@ def run(selenium, img, ref_sobel, ref_laplacian, ref_canny):
         laplacian = cv.Laplacian(gray, cv.CV_8U, ksize=3)
         canny = cv.Canny(src, 100, 255)
 
-        np.compare_(ref_sobel, sobel, cv.IMREAD_GRAYSCALE)
-        np.compare_(ref_laplacian, laplacian, cv.IMREAD_GRAYSCALE)
-        np.compare_(ref_canny, canny, cv.IMREAD_GRAYSCALE)
+        comp.obj(ref_sobel, sobel, cv.IMREAD_GRAYSCALE)
```
Comment (hoodmane) [inline packages/opencv-python/test_opencv_python.py:144]: This `comp.obj` notation isn't all that beautiful. I wonder if I should have just unpickled the `PyodideHandle` directly to the wrapped object...

Comment (ryanking13) [reply packages/opencv-python/test_opencv_python.py:144]: > This `comp.obj` notation isn't all that beautiful. I wonder if I should have just unpickled the `PyodideHandle` directly to the wrapped object...

That would be nice :)

Comment (ryanking13) [reply packages/opencv-python/test_opencv_python.py:144]: PyodideHandle works, thanks a lot!

But I found that it timeouts at the teardown step. It doesn't make test fail but it shows error such as:

```py
opencv-python/test_opencv_python.py::calib3d_chessboard[chrome] PASSED                                                                             [100%]
Exception ignored in: <function PyodideHandle.__del__ at 0x7f03977015a0>
Traceback (most recent call last):
  File "/src/.docker_home/.local/lib/python3.10/site-packages/pytest_pyodide/decorator.py", line 96, in __del__
    self.selenium.run_js(
  File "/src/.docker_home/.local/lib/python3.10/site-packages/pytest_pyodide/runner.py", line 249, in run_js
    return self.run_js_inner(code, check_code)
  File "/src/.docker_home/.local/lib/python3.10/site-packages/pytest_pyodide/runner.py", line 349, in run_js_inner
    retval = self.driver.execute_async_script(wrapper % (code, check_code))
  File "/usr/local/lib/python3.10/site-packages/selenium/webdriver/remote/webdriver.py", line 900, in execute_async_script
    return self.execute(command, {
  ...
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=41453): Max retries exceeded with url: /session/4370d6910ba26ddd6567840b97dfc598/execute/async (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f039505b5e0>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

My guess is selenium remote driver dies before deleting handle, do you have any ideas?

Comment (ryanking13) [reply packages/opencv-python/test_opencv_python.py:144]: Oh, never mind. There was a typo.

**Comment by ryanking13** [comment]
Opened a separate PR for test failure unrelated to this: #3202

File: packages/matplotlib/test_matplotlib.py [hunk 3]
```diff
@@ -345,36 +345,31 @@ def test_draw_text_rotated(selenium_standalone):
...
+        REFERENCE_IMAGES_PATH / f"canvas-text-rotated-{selenium.browser}.png"
+    ).read_bytes()
+    handle = compare_func_handle(selenium)
+    patch_font_loading_and_dpi(selenium, handle)
+    run(selenium, handle, ref)
 
 
 @matplotlib_test_decorator
+@pytest.mark.xfail(
+    reason="TODO: pytest_pyodide.pyodide.JsException: InvalidStateError: An attempt was made to use an object that is not, or is no longer, usable"
```
Comment (ryanking13) [inline packages/matplotlib/test_matplotlib.py:361]: This test started to fail, I guess it is also related to #3205.

**Comment by ryanking13** [comment]
The cmdline-runner test failure is not related to this PR. Opened a issue about it (#3207).

File: packages/matplotlib/test_matplotlib.py [hunk 4]
```diff
@@ -69,74 +33,103 @@ def save_canvas_data(selenium, output_path):
...
+            while not font_loaded[0]:  # wait until font is loading
+                await asyncio.sleep(0.2)
+
+            canvas_data = plt.gcf().canvas.get_pixel_data()
+            ref_data = np.asarray(Image.open(io.BytesIO(ref)))
+
+            deviation = np.mean(np.abs(canvas_data - ref_data))
+            assert float(deviation) == 0.0
+
+        return PyodideHandle({"compare": compare, "font-loaded": font_loaded})
```
Comment (hoodmane) [inline packages/matplotlib/test_matplotlib.py:86]: How about
```suggestion
        class result:
            compare = compare
            font_loaded = font_loaded
        return PyodideHandle(result)
```
and then elsewhere use `handle.font_loaded` and `handle.compare`.

File: packages/opencv-python/test_opencv_python.py [hunk 5]
```diff
@@ -227,34 +214,37 @@ def test_video_optical_flow(selenium):
...
         ret, old_frame = cap.read()
         assert ret
 
         old_gray = cv.cvtColor(old_frame, cv.COLOR_BGR2GRAY)
-        p0 = cv.goodFeaturesToTrack(old_gray, mask = None, **feature_params)
+        p0 = cv.goodFeaturesToTrack(old_gray, mask=None, **feature_params)
         # Create a mask image for drawing purposes
         mask = np.zeros_like(old_frame)
-        while(1):
+        while 1:
```
Comment (hoodmane) [inline packages/opencv-python/test_opencv_python.py:235]:
```suggestion
        while True:
```

**Review by hoodmane — APPROVED** [review]
Generally looks good to me.

File: packages/matplotlib/test_matplotlib.py (hunk 4 above)
Comment (ryanking13) [reply packages/matplotlib/test_matplotlib.py:86]: Sounds good, thanks!

File: packages/opencv-python/test_opencv_python.py (hunk 5 above)
Comment (ryanking13) [reply packages/opencv-python/test_opencv_python.py:235]: Thanks! IIRC I copied all examples from [opencv-python docs](https://docs.opencv.org/4.x/d6/d00/tutorial_py_root.html) without thinking 😂

**Comment by ryanking13** [comment]
Thanks again @hoodmane! PyodideHandle is really cool.

**Pull request merged**